packages/wrangler/src/d1/execute.tsx

```diff
@@ -104,50 +104,64 @@ export const Handler = async (args: HandlerOptions): Promise<void> => {
 		return logger.error(`Error: can't provide both --command and --file.`);
 
 	const isInteractive = process.stdout.isTTY;
-	const response: QueryResult[] | null = await executeSql({
-		local,
-		config,
-		name: database,
-		shouldPrompt: isInteractive && !yes,
-		persistTo,
-		file,
-		command,
-		json,
-		preview,
-		batchSize,
-	});
+	try {
+		const response: QueryResult[] | null = await executeSql({
+			local,
+			config,
+			name: database,
+			shouldPrompt: isInteractive && !yes,
+			persistTo,
+			file,
+			command,
+			json,
+			preview,
+			batchSize,
+		});
 
-	// Early exit if prompt rejected
-	if (!response) return;
+		// Early exit if prompt rejected
+		if (!response) return;
 
-	if (isInteractive && !json) {
-		// Render table if single result
-		logger.log(
-			renderToString(
-				<Static items={response}>
-					{(result) => {
-						// batch results
-						if (!Array.isArray(result)) {
-							const { results, query } = result;
+		if (isInteractive && !json) {
+			// Render table if single result
+			logger.log(
+				renderToString(
+					<Static items={response}>
+						{(result) => {
+							// batch results
+							if (!Array.isArray(result)) {
+								const { results, query } = result;
 
-							if (Array.isArray(results) && results.length > 0) {
-								const shortQuery = shorten(query, 48);
-								return (
-									<>
-										{shortQuery ? <Text dimColor>{shortQuery}</Text> : null}
-										<Table data={results}></Table>
-									</>
-								);
+								if (Array.isArray(results) && results.length > 0) {
+									const shortQuery = shorten(query, 48);
+									return (
+										<>
+											{shortQuery ? <Text dimColor>{shortQuery}</Text> : null}
+											<Table data={results}></Table>
+										</>
+									);
+								}
 							}
-						}
-					}}
-				</Static>
-			)
-		);
-	} else {
-		// set loggerLevel back to what it was before to actually output the JSON in stdout
-		logger.loggerLevel = existingLogLevel;
-		logger.log(JSON.stringify(response, null, 2));
+						}}
+					</Static>
+				)
+			);
+		} else {
+			// set loggerLevel back to what it was before to actually output the JSON in stdout
+			logger.loggerLevel = existingLogLevel;
+			logger.log(JSON.stringify(response, null, 2));
+		}
+	} catch (error) {
+		if (json) {
+			logger.loggerLevel = existingLogLevel;
+			const messageToDisplay =
+				(error as Error).name === "APIError"
+					? error
+					: { text: (error as Error).message };
+			logger.log(JSON.stringify({ error: messageToDisplay }, null, 2));
+			process.exit(1);
```

Comment:
This no longer closes wrangler with exitCode = 1 without doing it explicitly (and breaking tests), any tips on what you'd do here @petebacondarwin?

Reply:
Our approach is to throw a `new FatalError()` which can bubble up to the top level exit the process with a non-zero code.

Reply:
Is the output of that clean, though?

Reply:
We want wrangler to output JSON-friendly errors here, like:
```
{
	"error": {
		"text": "A request to the Cloudflare API (/accounts/xxxx/d1/database/xxxxxxx/query) failed.",
		"notes": [
			{
				"text": "no such column: asdf at offset 7 [code: 7500]"
			}
		],
		"kind": "error",
		"name": "APIError",
		"code": 7500
	}
}
```

Reply:
Ah nope:
```

✘ [ERROR] {

    "error": {
      "text": "A request to the Cloudflare API
  (/accounts/xxx/d1/database/xxx/query)
  failed.",
      "notes": [
        {
          "text": "no such column: asdf at offset 7 [code: 7500]"
        }
      ],
      "kind": "error",
      "name": "APIError",
      "code": 7500
    }
  }


If you think this is a bug then please create an issue at https://github.com/cloudflare/workers-sdk/issues/new/choose
🪵  Logs were written to "/Users/rozenmd/.wrangler/logs/wrangler-2024-01-31_10-44-47_583.log"
```

Reply:
😢 

OK so we really need to do a more holistic fix for `--json` output across the whole of Wrangler.
But in the meantime, how about you add a property to `FatalError` that tells us not to log the error?

Reply:
Basically check it here: https://github.com/cloudflare/workers-sdk/blob/main/packages/wrangler/src/index.ts#L822

Reply:
Gotcha, cheers @petebacondarwin !